### ISSUE TITLE
Recruit dialog redraw error

### ISSUE THREAD

**Issue by jostephd** [wesnoth/wesnoth#3942]
Apply this patch:

```
diff --git a/data/core/units/humans/Mage_of_Light.cfg b/data/core/units/humans/Mage_of_Light.cfg
index 74c7a268fad..e1f96074a90 100644
--- a/data/core/units/humans/Mage_of_Light.cfg
+++ b/data/core/units/humans/Mage_of_Light.cfg
@@ -55,7 +55,7 @@ Following a strict code of piety and honor, these men and women work tirelessly
         name=lightbeam
         description=_"lightbeam"	# wmllint: no spellcheck (until name->id)
         type=arcane
-        range=ranged
+        range=foobar
         [specials]
             {WEAPON_SPECIAL_MAGICAL}
         [/specials]
```

Start `./wesnoth --all-translations -L la_IT -t`. Press Ctrl+R to open the recruit dialog, type 'mag' and use the down arrow key to select the Mage of Light.

Result: the dialog is redrawn shorter than it was, but the portion between the old and new top/bottom margins isn't erased:

![screenshot_2019-02-24_13-38-07](https://user-images.githubusercontent.com/24784687/53299987-84929100-3839-11e9-8b50-9d92758012f6.png)

This is what I reported on https://github.com/wesnoth/wesnoth/pull/3787#issuecomment-447928428

The log has:

```
error gui/layout: grid [] place: Failed to place a grid, we have 194,20 space but we need 211,20 space. This happened at a grid with the id '' in a 'N4gui214tree_view_nodeE' with the id '' in a 'N4gui29tree_viewE' with the id 'type_details' in a 'N4gui24gridE' with the id '' in a 'N4gui217unit_preview_paneE' with the id 'recruit_details' in a 'N4gui24gridE' with the id '' in a 'N4gui24gridE' with the id '_window_content_grid' in a 'N4gui24gridE' with the id '_content_grid' in a 'N4gui215scrollbar_panelE' with the id '' in a 'N4gui24gridE' with the id '' in a 'N4gui26windowE' with the id 'unit_recruit'.
```

**Comment by stevecotton** [comment]
If something on the left-hand pane causes the dialog to resize, then this can happen. With HDPI fonts, both the recruit and recall dialogs can resize when moving the selection to a unit with a longer [unit]name - generally this results in the dialog getting wider but keeping the same height, because the HDPI fonts have already made the dialog use its the maximum height.

A fix to a separate issue (that the dialog was too short on HDPI) was applied to the recall dialog in f3a472ab963a5ff04b738ccb90de1c151039c1b6. I tried applying the same change to the recruit and advance dialogs, as they are currently too short (in 1.15.9+dev), and it makes this resize-to-shorter issue easily reproducible  - just advance an Elvish Shaman on an HDPI screen.

Stats of the screen that I'm reproducing this on are in #5544, and I think this issue should be used to fix all three dialogs. The ideal fix would probably be to lay out based on all possible options, but that's not going to be a simple change.

**Comment by mesilliac** [comment]
This looks like the same problem as #4877 which is confirmed fixed. Reopen if it reoccurs after 1.17.6.

**Comment by Wedge009** [comment]
In or after 1.17.6. Or just test 1.17.5+dev now.

**Comment by Descacharrado** [comment]
It does happen

**Comment by mesilliac** [comment]
How do you reproduce it? For me the dialog doesn't resize no matter what i do.

**Comment by Descacharrado** [comment]
I recorded the process described above:
https://youtu.be/UI1nM1b5n40

**Comment by Wedge009** [comment]
Strange, testing against 91983c8c875 I don't observe the issue with English (GB) in either Windows or Linux, with Mage of Light's range set to the proper 'ranged' value or a dummy 'fred' value.

_However_, I do observe the following in French language _without_ any editing of the unit definition. So maybe there is something in the length of strings that triggers the issue, somehow. Observe the left edge of the dialogue - the height of that left-over draw is where the previous dialogue (before the unit filtering) was.

![1 17 5+dev](https://user-images.githubusercontent.com/13679322/179327522-7dd1ffb6-fa5e-4c32-848b-207b15d98589.png)

**Comment by mesilliac** [comment]
I also found that the list of units must be small enough to have no scrollbar. If there is a scrollbar it seems not to resize.

I looked into the code and it's a bit complex... there may be an easy fix somewhere but it wasn't obvious at first glance.

**Comment by mesilliac** [comment]
FTR the easy fix would be calling `queue_redraw` before resizing, or `queue_redraw(old region)` after resizing. It's just a matter of knowing what is doing the resizing.

It might become more obvious if things get moved to a system more compatible with the new draw manager, but currently everything makes layout changes whenever it feels like it so it's difficult to track things like this down.

**Comment by grz0** [comment]
I noticed a similar issue with Statistics dialogue. I reported in #4695 that the statistics dialogue has a variable width, but then it was redrawn correctly. Now I observe artifacts like the below:
![image](https://user-images.githubusercontent.com/3605541/182499055-91dae628-c34a-415a-a884-ab77c364a9ed.png)

**Comment by Wedge009** [comment]
I'm not able to replicate this any more at e24f60fd432 (after #7974) - if anyone can, please re-open.
